PodSecurity admission

liggitt opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Initial implementation of the core functionality for the PodSecurity feature

#### Special notes for your reviewer:

Best reviewed commit-by-commit. Commits are ordered by package dependency order.

Remaining follow-up tasks tracked in https://github.com/kubernetes/kubernetes/issues/103192

Fixes https://github.com/kubernetes/kubernetes/issues/103215

Fixes https://github.com/kubernetes/kubernetes/issues/103197


#### Does this PR introduce a user-facing change?
```release-note
kube-apiserver: the alpha PodSecurity feature can be enabled by passing `--feature-gates=PodSecurity=true`, and enables controlling allowed pods using namespace labels. See https://git.k8s.io/enhancements/keps/sig-auth/2579-psp-replacement for more details.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-auth/2579-psp-replacement
```

/sig auth security

tallclair committed:
PodSecurity: api: runtime API

Label keys, values, and parsing helper functions

Co-authored-by: Jordan Liggitt <liggitt@google.com>

liggitt committed:
PodSecurity: admission/api: configuration API

Admission configuration:
- user, namespace, runtimeclass exemptions
- default policy levels and versions
- defaulting
- load and serialization helpers

Co-authored-by: Tim Allclair <timallclair@gmail.com>

liggitt committed:
PodSecurity: metrics: stub interface

Co-authored-by: Tim Allclair <timallclair@gmail.com>

tallclair committed:
PodSecurity: policy: registry

Co-authored-by: Jordan Liggitt <liggitt@google.com>

liggitt committed:
PodSecurity: test: framework

tallclair committed:
PodSecurity: admission: admission library

Co-authored-by: Jordan Liggitt <liggitt@google.com>

liggitt committed:
PodSecurity: pkg/features: feature gate

liggitt committed:
PodSecurity: kube-apiserver: admission wiring

liggitt committed:
PodSecurity: kube-apiserver integration test

liggitt committed:
PodSecurity: check: selinux

Co-authored-by: Tim Allclair <timallclair@gmail.com>

liggitt committed:
PodSecurity: check: runAsNonRoot

liggitt committed:
PodSecurity: check: allowPrivilegeEscalation

liggitt committed:
PodSecurity: test: generate fixture data

liggitt committed:
PodSecurity: vendor: generated files

Co-authored-by: Tim Allclair <timallclair@gmail.com>

fedebongio commented:
/remove-sig api-machinery

liggitt committed:
PodSecurity: update dependencies

caesarxuchao commented:
/triage accepted

deads2k commented on staging/src/k8s.io/pod-security-admission/api/helpers.go:
```diff
@@ -0,0 +1,202 @@
+		p.Warn.Level, err = LevelToEvaluate(level)
+		errs = appendErr(errs, err, "Warn.Level")
+		if err != nil {
+			p.Audit.Level = LevelPrivileged // Fail open for warn.
```
p.Warn.Level ?

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/api/register.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+// GroupName is the group name use in this package
+const GroupName = "apiserver.config.k8s.io"
```
Interesting that this group gets split across multiple repos.  Not strictly a problem.  This one is specific to a kube-apiserver versus generic apiservers though.

liggitt replied:
I was planning for the same config file to be able to be consumed by the standalone webhook

deads2k replied:
> I was planning for the same config file to be able to be consumed by the standalone webhook

The API is still kube specific, right?  As in, the configuration here only makes sense when applied to a level of the kube-apiserver pods. Versus the existing apiserver.config.k8s.io which makes sense when applied to any generic apiserver.

Not a point I feel very strongly about, but it did surprise me a bit.

liggitt replied:
huh... I could have sworn that we had other kubeapiserver admission config types using this group, but... I couldn't find any

liggitt replied:
hmm... I don't like any of the prior art for group names for admission plugins here:

Plugin-specific groups (doesn't end in `.config.k8s.io`, and a group per plugin seems excessive):
* eventratelimit.admission.k8s.io
* podtolerationrestriction.admission.k8s.io
* resourcequota.admission.k8s.io

Current group (as you point out, a pod-oriented admission plugin is specific to kube-apiserver):
* apiserver.config.k8s.io

liggitt replied:
went with `pod-security.config.admission.k8s.io`

* `pod-security`: matches format of the label keys for the feature
* `admission`: since this is an `admission` component that may or may not be inside an API server
* `config.k8s.io`: matches convention for configuration file types vs REST API types

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/api/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,48 @@
+
+func SetDefaults_PodSecurityDefaults(obj *PodSecurityDefaults) {
+	if len(obj.Enforce) == 0 {
+		obj.Enforce = string(api.LevelPrivileged)
```
I don't care for defaulting open like this.  Can we just leave it empty and count on the configuration to apply values instead of the defaulting and converting process?

liggitt replied:
discussed in review... we either default here or default deeper in... leaving config defaulting with a valid set of values is easier to reason about

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/api/validation/validation.go:
```diff
@@ -0,0 +1,28 @@
+
+func ValidatePodSecurityConfiguration(configuration *api.PodSecurityConfiguration) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// TODO: validate default levels and versions
```
filled in later?

liggitt replied:
yeah, tracked in https://github.com/kubernetes/kubernetes/issues/103193

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/checks.go:
```diff
@@ -0,0 +1,141 @@
+	Allowed bool
+	// ForbiddenReason should only be set if Allowed is false.
+	// ForbiddenReason should be as succinct as possible and is always output.
+	ForbiddenReason string
```
does this field mirror condition Reasons as CamelCase and machine parseable?

liggitt replied:
no, this is a "short version" of the error (e.g. `allowPrivilegeEscalation != false`, `forbidden seLinuxOptions`)

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/checks.go:
```diff
@@ -0,0 +1,141 @@
+type CheckResult struct {
+	// Allowed indicates if the check allowed the pod.
+	Allowed bool
+	// ForbiddenReason should only be set if Allowed is false.
```
if allowed is false, does this have to be set?

liggitt replied:
it *should* be, it's not enforced in tests now but could be

liggitt replied:
added check in tests

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/checks.go:
```diff
@@ -0,0 +1,141 @@
+	for _, check := range checks {
+		r := check.CheckPod(podMetadata, podSpec)
+		if !r.Allowed {
+			reasons = append(reasons, r.ForbiddenReason)
```
empties look like they'll produce weird output.

liggitt replied:
can fail if we encounter empty strings (or do ***MISSING*** or something and check for that in tests)

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/registry.go:
```diff
@@ -0,0 +1,185 @@
+	}
+}
+
+func (r *CheckRegistry) CheckForIDAndVersion(id string, version api.Version) (Check, error) {
```
This method may give me a baseline check or a restricted check?  That surprises me a little bit, I guess I'll see how it's used.

liggitt replied:
this is unused now, I think it was intended to allow access to a specific check for consumers you want to override/wrap

maybe remove until needed or replace with direct accessor to specific checks (GetSELinuxCheck(), etc)

liggitt replied:
removed in reworked registry approach

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/registry.go:
```diff
@@ -0,0 +1,185 @@
+	return defaultRegistry
+}
+
+var defaultRegistry = NewCheckRegistry()
```
I prefer not having globals.  It seems like we could handle this similar to schemes where an instance is instantiated.

liggitt replied:
will rework to register a list of default checks and make a InstallChecks(registry) method to install those checks in a particular registry instance

liggitt replied:
done

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+	} else {
+		if p, err := admissionapi.ToPolicy(a.Configuration.Defaults); err != nil {
+			return err
+		} else if !reflect.DeepEqual(p, a.defaultPolicy) {
```
note to self, track this down.

liggitt replied:
will clean up, this is trying to make sure Complete() was run and the results match (to avoid mutating in validation)

liggitt replied:
after frowning at this a bit, I still don't want to mutate in validation, or after calling validation... this feels awkward, but I might leave it for now (externally, calling Complete, then Validate feels ok)

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+func (a *Admission) ValidatePod(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	// short-circuit on subresources other than ephemeralcontainers
+	if attrs.GetSubresource() != "" && attrs.GetSubresource() != "ephemeralcontainers" {
+		return allowedResponse()
```
This allows by default instead of fully enumerating and denying unknown?  If this is used as a library by a webhook which may skew levels, I would expect enumeration and deny-by-default.

liggitt replied:
yeah, enumerating seems fine

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+	}
+}
+
+// isSignificantPodUpdate determines whether a pod update should trigger a policy evaluation.
```
Is there a shortcut link to which pod fields are mutable to help with reasoning about whether we got everything?

liggitt replied:
only API validation :-/

https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/apis/core/validation/validation.go#L3947-L3949

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+// The enforce policy is only checked if enforce=true.
+func (a *Admission) CheckPod(ctx context.Context, podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec, enforce bool) admissionv1.AdmissionResponse {
+	// short-circuit on exempt runtimeclass
+	if a.exemptRuntimeClass(podSpec.RuntimeClassName) {
```
I haven't read yet, but I'm betting this is here because namespace admission reuses this method.  There's a unit test ensuring someone doesn't optimize this into the pod evaluation method?

liggitt replied:
Both pod-controller and pod validation route here. Namespace evaluation was missing this. Added and added a unit test.

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+		}
+		if !isSignificantPodUpdate(pod, oldPod) {
+			// Nothing we care about changed, so always allow the update.
+			enforce = false
```
if nothing significant changed, why don't we return right here?  Audit messages on insignificant pod updates seems a little odd to me.

liggitt replied:
I sort of agree, though this matches what was [specified](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement#updates):

> Update requests to Pods and PodTemplate resources will reevaluate the full object against audit & warn policies, independent of which fields are being modified.

Tim, hitting warn/audit on controller updates makes sense to me... not sure I see the value on a pod update of fields we don't care about

deads2k replied:
> I sort of agree, though this matches what was [specified](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement#updates):
> 
> > Update requests to Pods and PodTemplate resources will reevaluate the full object against audit & warn policies, independent of which fields are being modified.
> 
> Tim, hitting warn/audit on controller updates makes sense to me... not sure I see the value on a pod update of fields we don't care about

I'd like a more complete explanation of the benefit we're getting by auditing and warning on field updates that aren't inspected by this admission plugin.

tallclair replied:
We definitely discussed this during the design phase, but I can't remember what the arguments were for always doing audit & warn. I'm OK with skipping those entirely.

liggitt replied:
changed to `return allowedReponse()` at this point

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+		response.Warnings = append(response.Warnings, fmt.Sprintf("Pod violates PodSecurity profile %s: %s", policy.Warn.String(), result.ForbiddenDetail()))
+	}
+
+	response.AuditAnnotations = auditAnnotations
```
nit: move up to the block where you set these?

liggitt replied:
we set annotations in a couple different blocks (an error annotation if the namespace labels didn't parse right)... doing it at the end seems easier to reason about to me

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+	PodSpecExtractor PodSpecExtractor
+
+	// API connections
+	NamespaceGetter NamespaceGetter
```
note to self, come back to ensure this punches through on misses, possibly using the LRU cache approach we used elsewhere

liggitt replied:
the implementation we're using does:

https://github.com/kubernetes/kubernetes/blob/eaa0a432f56e732473dc79fdcb370d50de70577d/staging/src/k8s.io/pod-security-admission/admission/namespace.go#L42-L50

deads2k commented on staging/src/k8s.io/pod-security-admission/admission/admission.go:
```diff
@@ -0,0 +1,537 @@
+	return policy.AggregateCheckPod(checks, podMetadata, podSpec), nil
+}
+
+func (a *Admission) CheckPodsInNamespace(ctx context.Context, namespace string, enforce api.LevelVersion) []string {
```
note to self, you didn't review this because you're trying to find the check registry before 10am

liggitt commented:
sigh... alpha e2e was failing because of https://github.com/kubernetes/kubernetes/issues/88282 ... all the test fixtures set pod.namespace, but in reality, objects may or may not have a namespace when they hit admission:

> I0626 04:49:46.861545       9 request.go:936] Error in request: resource name may not be empty
> E0626 04:49:46.861628       9 admission.go:318] "failed to fetch pod namespace" err="resource name may not be empty" namespace=""

good incentive to pick https://github.com/kubernetes/kubernetes/pull/94637 back up

liggitt committed:
PodSecurity: check: addCapabilities

liggitt commented:
Resolved most comments. Changes are in fixup commits. Remaining unresolved comments:
* API group for config file - https://github.com/kubernetes/kubernetes/pull/103099#discussion_r658744318
* question about warning/auditing on pod updates which don't touch any relevant fields - https://github.com/kubernetes/kubernetes/pull/103099#discussion_r658775723

deads2k commented on staging/src/k8s.io/pod-security-admission/policy/check_addCapabilities.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+func addCapabilities_1_0(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec) CheckResult {
+	forbiddenCapabilities := sets.NewString()
```
Minor, I think it would be nice to know which containers violated.

liggitt replied:
added container names to detail

deads2k commented:
for an initial cut, this lgtm

/lgtm
/hold  for @tallclair and @enj if they wish to review

tallclair commented:
I'm a co-author, so implicit LGTM from me.

liggitt commented:
> Remaining unresolved comments:
> 
> * API group for config file - [PodSecurity admission #103099 (comment)](https://github.com/kubernetes/kubernetes/pull/103099#discussion_r658744318)

Settled on `pod-security.admission.config.k8s.io`

> * question about warning/auditing on pod updates which don't touch any relevant fields - [PodSecurity admission #103099 (comment)](https://github.com/kubernetes/kubernetes/pull/103099#discussion_r658775723)

Resolved to skip evaluations for irrelevant pod updates

liggitt commented:
Dropped the "dropCapabilities" check from the initial PR because we're still debating whether to require dropping NET_RAW or ALL in 1.22 (https://docs.google.com/document/d/1Z26yS26z5saILHzj9nSvlbGOBJP_6c182Fl-vAi8AQY/edit#heading=h.jmjfcor83cb)

Remaining commits squashed and rebased on master

liggitt commented:
/hold cancel

liggitt commented:
readding based on https://github.com/kubernetes/kubernetes/pull/103099#issuecomment-869840801

checked offline, @enj indicated his review was non-blocking

pierluigilenoci commented:
Love it ❤️
